src/test/java/com/microsoft/sqlserver/jdbc/AlwaysEncrypted/MultiUserAKVTest.java

```diff
@@ -0,0 +1,643 @@
+                    break;
+                case "CEK":
+                    dropCEK(objectName, stmt);
+                    break;
+                case "CMK":
+                    dropCMK(objectName, stmt);
+                    break;
+                default:
+                    break;
+            }
+        } catch(Exception ex) {
+            fail(ex.getMessage());
+        }
+    }
+
+    private boolean cekCacheContainsKey(byte[] encryptedKey, SQLServerColumnEncryptionAzureKeyVaultProvider provider) throws Exception {
+        assertFalse(null == encryptedKey || 0 == encryptedKey.length);        
+
+        String encryptedCEKHexString = byteToHexDisplayString(encryptedKey);    
+
+        Field cekCacheField = provider.getClass().getDeclaredField(cekCacheName);
+        cekCacheField.setAccessible(true);
+
+        Object fieldValue = cekCacheField.get(provider);
+        Method method = fieldValue.getClass().getDeclaredMethod("contains", Object.class);
+        method.setAccessible(true);
+
+        boolean result = (boolean)method.invoke(fieldValue, encryptedCEKHexString);
+        
+        return result;
+    }
+
+    private byte[] signColumnMasterKeyMetadata(SQLServerColumnEncryptionAzureKeyVaultProvider provider, String masterKeyPath, boolean allowEnclaveComputations) throws Exception {
+
+        MessageDigest md = MessageDigest.getInstance("SHA-256");
+        md.update(Constants.AZURE_KEY_VAULT_NAME.toLowerCase().getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        md.update(masterKeyPath.toLowerCase().getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        
+        if (allowEnclaveComputations) {
+            md.update("true".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        } else {
+            md.update("false".getBytes(java.nio.charset.StandardCharsets.UTF_16LE));
+        }
+
+        byte[] dataToSign = md.digest();
+        if (null == dataToSign) {
+            fail("data to sign is null or empty.");
+        }
+
+        assertTrue(dataToSign.length > 0);
+
+        Method method = provider.getClass().getDeclaredMethod("AzureKeyVaultSignHashedData", byte[].class, String.class);
+        method.setAccessible(true);
+
+        byte[] signature = (byte[]) method.invoke(provider, dataToSign, masterKeyPath);
+
+        if (null == signature || 0 == signature.length) {
+            fail("Signature of column master key metadata is null or empty.");
+        }
+
+        return signature;
+    }
+    
+    private int getCacheSize(String methodName, SQLServerColumnEncryptionAzureKeyVaultProvider provider) throws Exception {
+        Method method = provider.getClass().getDeclaredMethod(methodName);
+        method.setAccessible(true);
+        
+        int count = (int)method.invoke(provider);
+
+        return count;     
+    }
+
+    private SQLServerColumnEncryptionAzureKeyVaultProvider createAKVProvider() throws Exception {

```

Comment:
why is this necessary? It's already done in setup() in AbstractTest?

Reply:
This is to create a new AKV provider instance that will have a fresh cache associated with it. I am trying to avoid the same akvProvider instance shared by other existing tests.